### ISSUE TITLE
Gradient Accumulation and AcceleratedScheduler

### ISSUE THREAD

**Issue by yocabon** [huggingface/accelerate#499]
Hi,

I think the learning rate scheduler is stepped every iteration, when it should be updated only when the optimizer was stepped.

Scheduler is skipped when: 
```
for opt in self.optimizers:
            if opt.step_was_skipped:
                return
```

but `opt.step_was_skipped` is the same as `self._is_overflow` and is not using self.gradient_state.

Also, [self.gradient_state](https://github.com/huggingface/accelerate/blob/main/src/accelerate/scheduler.py#L50) is not used at all in AcceleratedScheduler.

**Comment by sgugger** [comment]
No, because that would require the user having to divide the number of iterations of their scheduler by `gradient_accumulation_steps`. This way it's done automatically for the user.

**Comment by yocabon** [comment]
As I see it:

The sentence "Useful to avoid making a scheduler step too fast when step was skipped because of gradient accumulation" in the header of the AcceleratedScheduler class does not seem to match the implementation.

opt.step_was_skipped only corresponds to the overflow, and the learning rate scheduler is stepped every time step() is called on it (or am I not seeing the part where it's skipped ?)


Also in the example [here](https://github.com/huggingface/accelerate/blob/main/examples/by_feature/gradient_accumulation.py#L137), num_training_steps is divided by gradient_accumulation_steps

**Comment by muellerzr** [comment]
On that latter part, that's a mistake on my end and a PR to fix that will be coming shortly. 

Re: how does the optimizer know to be skipped, in the start of `step` and `zero_grad` is this conditional:
```python
if self.gradient_state.sync_gradients:
```
which gets modified based on if we should sync the gradients or not during gradient accumulation. And if `skip_sheduler_with_optimizer` is `True` (which it should always be unless a scheduler steps by epoch) then this is how the two link together and is skipped. 

For exactly where/how that's being set, see these lines in accelerator: https://github.com/huggingface/accelerate/blob/main/src/accelerate/accelerator.py#L393-L400

**Comment by yocabon** [comment]
Understood. It'd be great to also clarify : https://github.com/huggingface/accelerate/blob/main/src/accelerate/scheduler.py#L31
